### ISSUE TITLE
gRPC call hangs on Linux x64 in Azure Functions host

### ISSUE THREAD

**Issue by jskeet** [grpc/grpc#25208]
### What version of gRPC and what language are you using?

gRPC 2.34.1, using .NET Core 3.1.

### What operating system (Linux, Windows,...) and version?

Linux and Windows

### What runtime / compiler are you using (e.g. python version or version of gcc)

.NET Core SDK 3.1, with the azure-functions-core-tools-3 version 3.0.3160-1

### What did you do?

- Start an Azure function that uses gRPC to make a request, running under Linux.
- Invoke the function
- Function hangs

Zip file everything to reproduce using Docker: [repro.zip](https://github.com/grpc/grpc/files/5842408/repro.zip)

- `docker build -t grpc-repro`
- `docker run -p 7071:7071 -it grpc-repro`
- `func start` (and wait for the console to show that it's started)
- Browse to http://localhost:7071/api/TestFunction

This is the code that makes the request - it's not expected to *work*, just exercise gRPC:

```csharp
private static string MakeRequest()
{
    var marshaller = new Marshaller<string>(Encoding.UTF8.GetBytes, Encoding.UTF8.GetString);
    var method = new Method<string, string>(MethodType.Unary,
        "test-service", "test-method",
        marshaller, marshaller);
    var channel = new Channel("spanner.googleapis.com:443", ChannelCredentials.Insecure);
    var callInvoker = channel.CreateCallInvoker();
            
    return callInvoker.BlockingUnaryCall(method, "spanner.googleapis.com", new CallOptions(), "request");
}
```

### What did you expect to see?

An error of "StatusCode="Unavailable", Detail="failed to connect to all addresses"" due to making a bogus request (deliberately).

### What did you see instead?

The function hangs.

### Anything else we should know about your project / environment?

- The same code "fails correctly" when running in Windows in the Azure Functions host
- The same code "fails correctly" when running in Linux as a console application
- Using an async call doesn't help
- Real (correct) RPCs fail in the same way

**Comment by jtattermusch** [comment]
Looks like I can reproduce.

I ran `GRPC_VERBOSITY=debug func start` to get extra logs and this is the output:
```
D0129 12:03:24.526419 Grpc.Core.Internal.UnmanagedLibrary Attempting to load native library "/usr/lib/azure-functions-core-tools-3/libgrpc_csharp_ext.x64.so"
D0129 12:03:24.549782 Grpc.Core.Internal.NativeExtension gRPC native library loaded successfully.
D0129 12:03:24.550769 140687564564224 /var/local/git/grpc/src/core/lib/iomgr/ev_posix.cc:174: Using polling engine: epollex
D0129 12:03:24.550884 140687564564224 /var/local/git/grpc/src/core/ext/filters/client_channel/resolver/dns/c_ares/dns_resolver_ares.cc:503: Using ares dns resolver
I0129 12:03:24.573641 140687564564224 /var/local/git/grpc/src/core/lib/iomgr/socket_utils_common_posix.cc:387: Disabling AF_INET6 sockets because ::1 is not available.

Functions:

	TestFunction:  http://localhost:7071/api/TestFunction

For detailed output, run func with --verbose flag.
[2021-01-29T12:03:30.320Z] Host lock lease acquired by instance ID '00000000000000000000000012F43BA1'.
[2021-01-29T12:03:42.251Z] Executing 'TestFunction' (Reason='This function was programmatically called via the host APIs.', Id=75791b43-093e-4c6d-997d-e6bee19ae275)
D0129 12:03:42.290965 Grpc.Core.Internal.NativeExtension gRPC native library loaded successfully.
I0129 12:03:42.348189 140687538710272 /var/local/git/grpc/src/core/ext/filters/client_channel/subchannel.cc:1055: New connected subchannel at 0x7ff39c07ffd0 for subchannel 0x7ff3a8004de0
E0129 12:03:42.365032 Grpc.Core.Internal.AsyncCall`2 Exception occurred while invoking completion delegate. System.EntryPointNotFoundException: Unable to find an entry point named '?' in shared library 'grpc_csharp_ext.x64'.
   at Grpc.Core.Internal.NativeMethods.DllImportsFromSharedLib_x64.grpcsharp_batch_context_recv_status_on_client_error_string(BatchContextSafeHandle ctx)
   at Grpc.Core.Internal.BatchContextSafeHandle.GetReceivedStatusOnClient()
   at Grpc.Core.Internal.AsyncCall`2.UnaryCall(TRequest msg)
[2021-01-29T12:03:44.805Z] Executing 'TestFunction' (Reason='This function was programmatically called via the host APIs.', Id=f9a65949-b0bc-4641-8843-97bdc88913b1)
I0129 12:03:44.813852 140684930184960 /var/local/git/grpc/src/core/ext/filters/client_channel/subchannel.cc:1055: New connected subchannel at 0x7ff38401f920 for subchannel 0x7ff38400ca00
E0129 12:03:44.815697 Grpc.Core.Internal.AsyncCall`2 Exception occurred while invoking completion delegate. System.EntryPointNotFoundException: Unable to find an entry point named '?' in shared library 'grpc_csharp_ext.x64'.
   at Grpc.Core.Internal.NativeMethods.DllImportsFromSharedLib_x64.grpcsharp_batch_context_recv_status_on_client_error_string(BatchContextSafeHandle ctx)
   at Grpc.Core.Internal.BatchContextSafeHandle.GetReceivedStatusOnClient()
   at Grpc.Core.Internal.AsyncCall`2.UnaryCall(TRequest msg)
```

**Comment by jtattermusch** [comment]
The `System.EntryPointNotFoundException: Unable to find an entry point named '?' in shared library 'grpc_csharp_ext.x64'`  message is puzzling me

https://github.com/grpc/grpc/blob/86f93801e5d452cff40d6958d21a8eb211da0654/src/csharp/Grpc.Core/Internal/BatchContextSafeHandle.cs#L95

**Comment by jskeet** [comment]
I've now reproduced this on Windows, which should make it easier for the gRPC team to diagnose. Here's a zip file with everything:

[GrpcHang.zip](https://github.com/grpc/grpc/files/5994372/GrpcHang.zip)

The three projects are:

- GrpcLibrary: a class library targeting netstandard2.0 which just makes a garbage RPC to exercise gRPC
- ConsoleApp: a console app targeting multiple frameworks which has a dependency on GrpcLibrary and just calls the method to exercise gRPC
- ReflectionExecutor: a console app targeting multiple frameworks which doesn't have any dependencies, but which loads ConsoleApp via reflection (using whatever path is specified on the command line) and calls its Main method

There's a `run-test.sh` bash script to simplify testing. It takes three arguments:

- The TFM to use for ReflectionExecutor
- The TFM to use for ConsoleApp
- The version of Grpc.Core to use (by manipulating GrpcLibrary.csproj)

It sets GRPC_VERBOSITY to debug so we can see the same output as in the comment two above this one.

The most interesting pair of cases is probably to use netcoreapp3.1 for both TFMs, and Grpc.Core 2.33.1 (works) and 2.34.0 (hangs).

**Comment by jtattermusch** [comment]
i've played a bit more with the azure functions repro and at this point I'm pretty sure this issue is an unwanted side effect of https://github.com/grpc/grpc/pull/24744

I tried with Grpc.Core 2.33.1 (just before https://github.com/grpc/grpc/pull/24744 was added) and got a different error (and it didn't hang).
```
Error: System.IO.FileNotFoundException: Error loading native library. Not found in any of the possible locations: /app/bin/output/bin/libgrpc_csharp_ext.x64.so,/app/bin/output/bin/runtimes/linux/native/libgrpc_csharp_ext.x64.so,/app/bin/output/bin/../../runtimes/linux/native/libgrpc_csharp_ext.x64.so
   at Grpc.Core.Internal.UnmanagedLibrary.FirstValidLibraryPath(String[] libraryPathAlternatives)
   at Grpc.Core.Internal.UnmanagedLibrary..ctor(String[] libraryPathAlternatives)
   at Grpc.Core.Internal.NativeExtension.LoadUnmanagedLibrary()
   at Grpc.Core.Internal.NativeExtension.LoadNativeMethods()
   at Grpc.Core.Internal.NativeExtension..ctor()
   at Grpc.Core.Internal.NativeExtension.Get()
   at Grpc.Core.Internal.NativeMethods.Get()
   at Grpc.Core.GrpcEnvironment.GrpcNativeInit()
   at Grpc.Core.GrpcEnvironment..ctor()
   at Grpc.Core.GrpcEnvironment.AddRef()
   at Grpc.Core.Channel..ctor(String target, ChannelCredentials credentials, IEnumerable`1 options)
   at Grpc.Core.Channel..ctor(String target, ChannelCredentials credentials)
   at GrpcTest.TestFunction.MakeRequest() in /app/TestFunction.cs:line 35
   at GrpcTest.TestFunction.Run(HttpRequest req) in /app/TestFunction.cs:line 19
```

Basically, it seems that inside the azure functions docker, the grpc_csharp_ext shared library is in an unexpected place (will need to investigate further) and Grpc.Core isn't able to find it. In Grpc.Core 2.33.1 this ends up with an immediate and clear error, but  in 2.34.0+    the shared library in now loaded differently: 

https://github.com/grpc/grpc/pull/24744/files#diff-2c65a0fc0ff30ef55102682b95465c44e758e53921b0f36be1b427f1feb12e18R142

The difference is that the original mechanism is "eager" in a sense that if the shared library is not found, the error is immediately reported.   The standard "DllImport" way of importing shared libraries is lazy though - you won't get an error until you actually try to invoke one of the native functions. I guess that somehow doesn't play well with the gRPC initialization logic and the hang happens because  the shared library load is attempted too late (somewhere where it doesn't bubble up to the user) and some of the grpc logic just hangs.

I think the case from https://github.com/grpc/grpc/issues/25208#issuecomment-780400371 is similar in that the native shared library also can't be found. This time it might be because of the reflection  - the old native library loading mechanism tries to guess the location of grpc_csharp_ext  based on Grpc.Core.dll assembly's location (which happens to work well with the ReflectionExecutor appraoch). Meanwhile with the new native library loading logic,  the DllImport now relies on the .NET default native library loading logic. But the application context is setup around the ReflectionExecutor app (and searches for its "native dependencies'), it doesn't know anything about Grpc.Core's  runtime directory (which has the grpc_csharp_ext). So the native library load fails, and for the reasons described above this doesn't lead to an immediate crash, but hangs instead.
